### ISSUE TITLE
Improve `NetworkPolicy` for customer Namespaces

### ISSUE THREAD

**Issue by axisofentropy** [UffizziCloud/uffizzi_controller#25]
Right now we're defining `NetworkPolicy` Resources that block traffic coming into each Namespace from other Namespaces and Pod IP's. We must additional define "egress" rules to block a Pod's connection elsewhere. Right now customer workloads can make TCP connections to the k8s Master API, nodes, and other Namespaces that do not have a `NetworkPolicy` defined.

Customer workload Pods in Deployments specified by this controller should only be able to connect to itself, the load balancer, and The Internet.

**Comment by axisofentropy** [comment]
Our controller currently creates NetworkPolicy Resources for each customer Namespace that look like this.
This permits connections FROM each customer Pod TO all IP addresses EXCEPT for those within the 10.20.x.x range, which is all other Pods. There are exceptions for Pods within our load balancer's namespace (Three load balancer Namespaces are specified because it varies when installed by Terraform, Helm, or legacy.)

Outgoing "egress" connections are not specified so each Pod may connect to anything it can otherwise reach. This includes the k8s Master API, our controller, our k8s nodes, `kube-system` Pods, etc. Each customer Pod cannot connect to other customer Pods because they are still protected in turn by their own NetworkPolicy.

So in summary: our existing policy protects customer Pods from each other but nothing else.

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  labels:
    app.kubernetes.io/managed-by: uffizzi
  name: app-deployment-2129
  namespace: deployment-2129
spec:
  ingress:
  - from:
    - podSelector: {}
    - ipBlock:
        cidr: 0.0.0.0/0
        except:
        - 10.20.0.0/14
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: ingress-nginx
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: uffizzi-controller
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: ingress-nginx
  podSelector: {}
  policyTypes:
  - Ingress
```

**Comment by axisofentropy** [comment]
Here's an example new `NetworkPolicy` that I'm going to implement and test:
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  labels:
    app.kubernetes.io/managed-by: uffizzi
  name: app-deployment-2128
  namespace: deployment-2128
spec:
  egress:
  - to:
    - ipBlock:
        cidr: 0.0.0.0/0
        except:
        - 10.0.0.0/8
        - 172.16.0.0/12
        - 192.168.0.0/16
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: ingress-nginx
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: uffizzi-controller
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: ingress-nginx
  podSelector: {}
  policyTypes:
  - Ingress
  - Egress
```

**Comment by axisofentropy** [comment]
Changes include:

- Outgoing "egress" connections FROM each customer Pod TO the outside Internet are still permitted, but they're now blocked for ALL private IPv4 addresses. This means no more connections to other Pods of any kind,  nodes, the Master API, etc. These private IP ranges explained here: https://en.wikipedia.org/wiki/Private_network#Private_IPv4_addresses
- I've removed the "ingress" ipBlock exception because it's redundant. We don't want to permit any incoming connections at all except those from our load balancers.
- I'll probably also remove the redundant third load balancer namespace now that there we're not supporting any more "legacy" installations.

I've tested this new `NetworkPolicy` on `qa-gke` and confirmed all of the above statements.

**Comment by axisofentropy** [comment]
This PR is ready for review and merged to qa. https://github.com/UffizziCloud/uffizzi_controller/pull/27

**Comment by moklidia** [comment]
@axisofentropy Can you please guide @NealArw on how it can be tested?

**Comment by axisofentropy** [comment]
> @axisofentropy Can you please guide @NealArw on how it can be tested?

Yes sorry that's a great idea.

Testing
- Create a deployment on `qa-gke`; any container image that has a shell - `nginx` is fine.
- Confirm that you can connect to its public URL; this confirms that `ingress-nginx` can reach the Pod.
- Confirm that the deployment status is "green"; this confirms `uffizzi-controller` successfully checked the networking.
- Execute a shell within the deployed container: `kubectl exec -it <pod name> -n <namespace>`
- Install some networking tools: `apt-get update; apt-get install -y telnet curl hping3 net-tools ssh jq <anything else you need>`
- Identify some other Pods and `Services` on the Cluster whose IP address begins with `10.x.x.x`; confirm you cannot create TCP connections to those Pods.
- Identify the Kubernetes Master API endpoint from environment variables; confirm you cannot create TCP (or HTTPS) connections to this API.
- Confirm the Pod can connect to The Internet. `curl https://uffizzi.com` etc.
- Identify another customer workload Pod created **before** this change. Delete its `NetworkPolicy`. Confirm you cannot create TCP connections to it.
- Identify another customer workload Pod created **after** this change. Delete its `NetworkPolicy`. Confirm you cannot create TCP connections to it.
- Do the same for customer workload `Services`.
- Identify a Kubernetes Node IP address; confirm you cannot create SSH connections to it. `kubectl get nodes -o jsonpath='{.items[*].status.addresses[0].address}'`

Let me know if you have any questions! Great opportunity to learn more about Kubernetes networking.

**Comment by axisofentropy** [comment]
Sharing my testing:

Two Deployments:
- `nginx` from Docker Hub https://deployment-2252-default-adam.app.qa-gke.uffizzi.com/
- `docker-hello-world` from GitHub https://deployment-2253-default-adam.app.qa-gke.uffizzi.com/
Both URL's are accessible and both Deployments marked as happy green active; This means both the controller and load balancer namespaces could access the Pods.

```
adam@ephemeron:~$ gcloud container clusters get-credentials uffizzi-client --region us-central1 --project uffizzi-pro-qa-gke
```
![image](https://user-images.githubusercontent.com/66315/185200348-63cadb6f-6178-4b20-b5b9-63e02c27e913.png)

Delete the `NetworkPolicy` for the second Deployment for testing
```
adam@ephemeron:~$ kubectl delete networkpolicy app-deployment-2253 --namespace deployment-2253
networkpolicy.networking.k8s.io "app-deployment-2253" deleted
```

```
root@app-deployment-2252-f7747cc6c-22p4g:/# apt-get update; apt-get install -y telnet curl hping3 net-tools ssh jq
```
This command failed because this Pod could not contact its specified `nameserver` at `10.63.208.10`, so it needs another exception to contact the `kube-dns` Service on port `53`. I'll update the Pull Request with this.

**Comment by axisofentropy** [comment]
Resuming testing:

Can't connect to a node via SSH:
```
root@app-deployment-2256-78b7fcb888-xcn2n:/# ssh -v 10.128.0.102
( error after timeout )
```

Can connect to The Internet
```
root@app-deployment-2256-78b7fcb888-xcn2n:/# curl https://app.uffizzi.com
<!DOCTYPE html>
```

Cannot connect to the k8s Master API:
```
root@app-deployment-2256-78b7fcb888-xcn2n:/# curl https://$KUBERNETES_SERVICE_HOST
curl: (28) Failed to connect to 10.63.208.1 port 443: Connection timed out
```

and a few more tests. I think this is Doing What We Want now. It's ready for code review or staging or production or whatever's next in The Process.
